File: packages/plugin-sdk-py/src/aisoc_plugin_sdk/enricher.py
"""Enricher plugin base class."""

from __future__ import annotations

from abc import abstractmethod
from typing import Any

from pydantic import BaseModel, Field

from .plugin import AiSOCPlugin, PluginContext


class EnrichmentRequest(BaseModel):
    """Payload sent to an enricher plugin."""

    indicator_type: str = Field(
        ...,
        description="Type of indicator: ip | domain | url | hash | email",
    )
    indicator_value: str = Field(..., description="The indicator value to enrich")
    case_id: str | None = Field(None, description="Associated case ID, if any")
    metadata: dict[str, Any] = Field(default_factory=dict, description="Caller-supplied metadata")


class EnrichmentResult(BaseModel):
    """Result returned by an enricher plugin."""

    indicator_type: str
    indicator_value: str
    enrichments: dict[str, Any] = Field(
        default_factory=dict,
        description="Key-value enrichment data to merge into the indicator record",
    )
    tags: list[str] = Field(default_factory=list, description="Tags to apply to the indicator")
    malicious: bool | None = Field(None, description="True/False/None if enricher can determine")
    confidence: float | None = Field(
        None, ge=0.0, le=1.0, description="Confidence score 0–1, if applicable"
    )
    raw: dict[str, Any] = Field(
        default_factory=dict, description="Raw upstream API response for audit purposes"
    )


class EnricherPlugin(AiSOCPlugin):
    """Base class for enricher plugins."""

    @abstractmethod
    async def enrich(
        self, request: EnrichmentRequest, ctx: PluginContext
    ) -> EnrichmentResult:
        """Enrich an indicator and return structured results."""

File: packages/plugin-sdk-go/aisoc/enricher.go
package aisoc

import "context"

// IndicatorType classifies the type of security indicator being enriched.
type IndicatorType string

const (
	IndicatorIP     IndicatorType = "ip"
	IndicatorDomain IndicatorType = "domain"
	IndicatorURL    IndicatorType = "url"
	IndicatorHash   IndicatorType = "hash"
	IndicatorEmail  IndicatorType = "email"
)

// EnrichmentRequest is the payload sent to an enricher plugin.
type EnrichmentRequest struct {
	IndicatorType  IndicatorType  `json:"indicator_type"`
	IndicatorValue string         `json:"indicator_value"`
	CaseID         string         `json:"case_id,omitempty"`
	Metadata       map[string]any `json:"metadata,omitempty"`
}

// EnrichmentResult holds the enriched data returned by an enricher plugin.
type EnrichmentResult struct {
	IndicatorType  IndicatorType  `json:"indicator_type"`
	IndicatorValue string         `json:"indicator_value"`
	Enrichments    map[string]any `json:"enrichments,omitempty"`
	Tags           []string       `json:"tags,omitempty"`
	// Malicious is a tri-state: nil = unknown, true/false = determined.
	Malicious  *bool          `json:"malicious,omitempty"`
	Confidence *float64       `json:"confidence,omitempty"`
	Raw        map[string]any `json:"raw,omitempty"`
}

// Enricher is implemented by plugins that enrich security indicators.
type Enricher interface {
	Plugin
	Enrich(ctx context.Context, req EnrichmentRequest, pctx PluginContext) (EnrichmentResult, error)
}
